fmn/formatters.py

```diff
@@ -0,0 +1,426 @@
+            date_time = msg['timestamp']
+            date_time = datetime.datetime.fromtimestamp(date_time)
+            date_time = date_time.replace(tzinfo=pytz.utc).isoformat()
+        else:
+            date_time = datetime.datetime.utcnow().replace(tzinfo=pytz.utc).isoformat()
+
+        icon = fedmsg.meta.msg2icon(msg, **config.app_conf)
+        link = fedmsg.meta.msg2link(msg, **config.app_conf)
+        secondary_icon = fedmsg.meta.msg2secondary_icon(msg, **config.app_conf)
+        username = fedmsg.meta.msg2agent(msg, **config.app_conf)
+        title = fedmsg.meta.msg2title(msg, **config.app_conf)
+        subtitle = fedmsg.meta.msg2subtitle(msg, **config.app_conf)
+
+    if recipient['shorten_links']:
+        link = shorten(link)
+    event_link = ''
+    if title and link:
+        event_link = '<a href="' + link + '">' + title + '</a>'
+    user_link = ''
+    if username:
+        user_link = '<a href="/' + username.replace("@", "") + '">' + username + '</a>'
+
+    markup = ''
+    if user_link:
+        markup += user_link
+    if event_link:
+        markup += ' ' + event_link
+    if subtitle:
+        markup += ' ' + subtitle
+    markup = markup.strip()
+
+    output = {
+        'dom_id': dom_id,
+        'date_time': date_time,
+        'icon': icon,
+        'link': link,
+        'markup': markup,
+        'secondary_icon': secondary_icon
+    }
+
+    return json.dumps(output)
+
+
+def email(message, recipient):
+    """
+    Create a Python email Message that's ready for delivery.
+
+    This handles setting all the headers and formatting.
+
+    Args:
+        message (dict): The fedmsg to turn into an email.
+        recipient (dict): A dictionary containing (at a minimum) the
+            `email_address` (str) and `triggered_by_links` (bool) keys.
+
+    Returns:
+        str: The email as a unicode string.
+    """
+    from_address = config.app_conf['fmn.email.from_address']
+    topics = [message['topic']]
+    categories = [topics[0].split('.')[3]]
+    content = fedmsg.meta.msg2long_form(message, **config.app_conf) or u''
+    link = fedmsg.meta.msg2link(message, **config.app_conf) or u''
+    if link:
+        content += "\n\t" + link
+    subject = fedmsg.meta.msg2subtitle(message, **config.app_conf) or u''
+    usernames = fedmsg.meta.msg2usernames(message, **config.app_conf)
+    packages = fedmsg.meta.msg2packages(message, **config.app_conf)
+
+    email_message = email_module.Message.Message()
+    email_message.add_header('To', to_bytes(recipient['email address']))
+    email_message.add_header('From', to_bytes(from_address))
+    # Although this is a non-standard header and RFC 2076 discourages it, some
+    # old clients don't honour RFC 3834 and will auto-respond unless this is set.
+    email_message.add_header('Precendence', 'Bulk')
+    # Mark this mail as auto-generated so auto-responders don't respond; see RFC 3834
+    email_message.add_header('Auto-Submitted', 'auto-generated')
```

Comment:
Some of this code is repeated elsewhere in this patch. Is there an opportunity to share code here?

Reply:
Possibly. What should probably happen is that the other place should be refactored into this module. I'm inclined to do it in a separate PR, though.